python/ray/tests/test_generators.py

```diff
@@ -66,32 +73,243 @@ def generator(num_returns):
+            remote_generator_fn.options(num_returns=num_returns).remote(
+                num_returns + 1, store_in_plasma
+            )
         )
-        assert False
-    except ray.exceptions.RayTaskError as e:
-        assert isinstance(e.as_instanceof_cause(), ValueError)
-
-    # Check num_returns=1 case, should receive TypeError because generator
-    # cannot be pickled.
-    try:
-        ray.get(remote_generator_fn.remote(num_returns))
-        assert False
-    except ray.exceptions.RayTaskError as e:
-        assert isinstance(e.as_instanceof_cause(), TypeError)
+    else:
+        try:
+            ray.get(
+                remote_generator_fn.options(num_returns=num_returns).remote(
+                    num_returns + 1, store_in_plasma
+                )
+            )
+            assert False
+        except ray.exceptions.RayTaskError as e:
+            assert isinstance(e.as_instanceof_cause(), ValueError)
 
     # Check return values.
-    ray.get(
-        remote_generator_fn.options(num_returns=num_returns).remote(num_returns)
-    ) == list(range(num_returns))
+    [
+        x[0]
+        for x in ray.get(
+            remote_generator_fn.options(num_returns=num_returns).remote(
+                num_returns, store_in_plasma
+            )
+        )
+    ] == list(range(num_returns))
+    # Works for num_returns=1 if generator returns a single value.
+    assert (
+        ray.get(remote_generator_fn.options(num_returns=1).remote(1, store_in_plasma))[
+            0
+        ]
+        == 0
+    )
+
+
+@pytest.mark.parametrize("store_in_plasma", [False, True])
+def test_generator_errors(ray_start_regular, store_in_plasma):
+    @ray.remote(max_retries=0)
+    def generator(num_returns, store_in_plasma):
+        for i in range(num_returns - 2):
+            if store_in_plasma:
+                yield np.ones(1_000_000, dtype=np.int8) * i
+            else:
+                yield [i]
+        raise Exception("error")
+
+    ref1, ref2, ref3 = generator.options(num_returns=3).remote(3, store_in_plasma)
+    # TODO(swang): Currently there is a bug when a generator errors after
+    # already storing some values in plasma. The already stored values can
+    # be accessed and the error is lost.  Propagate the error correctly by
+    # replacing all successfully stored return values with the same error.
+    if not store_in_plasma:
+        with pytest.raises(ray.exceptions.RayTaskError):
+            ray.get(ref1)
+    with pytest.raises(ray.exceptions.RayTaskError):
+        ray.get(ref2)
+    with pytest.raises(ray.exceptions.RayTaskError):
+        ray.get(ref3)
+
+    dynamic_ref = generator.options(num_returns="dynamic").remote(3, store_in_plasma)
+    with pytest.raises(ray.exceptions.RayTaskError):
+        ray.get(dynamic_ref)
+
+
+@pytest.mark.parametrize("store_in_plasma", [False, True])
+def test_dynamic_generator(ray_start_regular, store_in_plasma):
```

Comment:
Can we test both task and actor task for dynamic generator?

Reply:
Good call, turns out it does not yet work on actors :D I'll leave it for a follow-up to unblock other work.